### PR TITLE
add table of locality attributes in refman

### PR DESCRIPTION
**Pull request by Villetaneuse**
Table of locality attributes behaviour in `modules.rst`.
TODO:
- [x] Check if refman looks good enough
- [x] Double-check that tables are right, possibly adding a test file

- table for modules: [here](https://coq.gitlabpages.inria.fr/-/coq/-/jobs/4641455/artifacts/_build/default/doc/refman-html/language/core/modules.html#summary-of-locality-attributes-in-a-module)
- table for sections: [here](https://coq.gitlabpages.inria.fr/-/coq/-/jobs/4641455/artifacts/_build/default/doc/refman-html/language/core/sections.html#summary-of-locality-attributes-in-a-section)

### CODE REVIEW

File: doc/sphinx/language/core/modules.rst [hunk 1]
```diff
@@ -1063,3 +1063,79 @@ while noting a few exceptional commands for which :attr:`local` and
...
+    - available with qualified name
+    - available with short name
+    - available with qualified name
+    - attribute not supported
+    - same as without attribute
+
+  * - :cmd:`Notation`
+    - not available outside
+    - available outside
+    - available with qualified name
```
Comment (SkySkimmer) [inline doc/sphinx/language/core/modules.rst:1090]: Isn't this Local Notation? Then it should be not available outside

File: doc/sphinx/language/core/modules.rst [hunk 2]
```diff
@@ -1063,3 +1063,79 @@ while noting a few exceptional commands for which :attr:`local` and
...
+    - attribute not supported
+    - same as without attribute
+
+  * - :cmd:`Abbreviation`
```
Comment (SkySkimmer) [inline doc/sphinx/language/core/modules.rst:1094]: not a real command?

Comment (SkySkimmer) [reply doc/sphinx/language/core/modules.rst:1094]: maybe intended to be
```suggestion
  * - :cmd:`Notation (abbreviation)`
```

File: doc/sphinx/language/core/modules.rst [hunk 3]
```diff
@@ -1063,3 +1063,79 @@ while noting a few exceptional commands for which :attr:`local` and
...
+    - always available outside
+
+  * - :cmd:`Canonical`
+    - not available outside
+    - available outside
+    - same as without attribute
+    - attribute not supported
+    - same as without attribute
+
+  * - :cmd:`Tactic`
```
Comment (SkySkimmer) [inline doc/sphinx/language/core/modules.rst:1122]: not a real command, maybe
```suggestion
  * - :cmd:`Ltac`
```

File: doc/sphinx/language/core/modules.rst [hunk 4]
```diff
@@ -1063,3 +1063,79 @@ while noting a few exceptional commands for which :attr:`local` and
...
       the user little control. We recommend using the :attr:`export`
       locality attribute where it is supported.
+
+The following table sums up the locality of vernacular commands in modules, when
+outside the module.
+
+.. list-table::
+  :header-rows: 1
+
+  * - :cmd:`Command`
```
Comment (SkySkimmer) [inline doc/sphinx/language/core/modules.rst:1073]: not a real command

File: doc/sphinx/language/core/modules.rst (hunk 1 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst:1090]: Yes, mistake on my part

File: doc/sphinx/language/core/modules.rst (hunk 4 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst]: Thanks, just changed it to ``Command``.

File: doc/sphinx/language/core/modules.rst (hunk 1 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst]: Changed

**Comment by Villetaneuse** [comment]
Added 4 test files and moved typing rules for modules at the end of the modules part in the refman.

File: doc/sphinx/language/core/modules.rst [hunk 5]
```diff
@@ -511,6 +511,249 @@ transparent constraint
 Notice that ``M`` is a correct body for the component ``M2`` since its ``T``
 component is ``nat`` as specified for ``M1.T``.
 
+.. _qualified-names:
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:514]: If you want to move a large block of text, please do that by itself in a separate commit or separate PR with a comment like "Move qualified names section" so it's easy to see the changes.  Also would help to eventually squash some of your commits so the changes are easy to pick out.

File: doc/sphinx/language/core/sections.rst [hunk 6]
```diff
@@ -131,6 +131,90 @@ usable outside the section as shown in this :ref:`example <section_local_declara
...
+The following table sums up the locality of vernacular commands in sections, when
+outside the section where it was entered.
+A similar table for :cmd:`Module` can be found
+:ref:`here <visibility-attributes-modules>`.
+
+.. list-table::
+  :header-rows: 1
+
+  * - ``Command``
+    - without attribute
```
Comment (jfehrle) [inline doc/sphinx/language/core/sections.rst:148]:
```suggestion
    - no attribute
```

File: doc/sphinx/language/core/sections.rst [hunk 7]
```diff
@@ -131,6 +131,90 @@ usable outside the section as shown in this :ref:`example <section_local_declara
...
+    - :attr:`local`
+    - :attr:`export`
+    - :attr:`global`
+
+  * - :cmd:`Definition`, :cmd:`Lemma`, :cmd:`Axiom`, ...
+    - same as :attr:`local`
+    - available
+    - attribute not supported
```
Comment (jfehrle) [inline doc/sphinx/language/core/sections.rst:156]:
```suggestion
    - available
    - not supported
```

Easier to read if the entries are more succinct.  Could you clarify the difference between "available" and "not supported" just before the table?  Does "not supported" mean you get an error message or that it's ignored?  (Ignored seems like it confuse the user.)  You could use just "✔" and "❌" to make it even easier to read.

File: doc/sphinx/language/core/sections.rst [hunk 8]
```diff
@@ -131,6 +131,90 @@ usable outside the section as shown in this :ref:`example <section_local_declara
...
+
+  * - ``Hints`` (and :cmd:`Instance`)
+    - same as :attr:`local`
+    - not available
+    - attribute not supported
+    - attribute not supported
+
+  * - :cmd:`Set` or :cmd:`Unset` a flag
+    - same as :attr:`local`
+    - no effect
```
Comment (jfehrle) [inline doc/sphinx/language/core/sections.rst:179]: Should there be a warning that there is no effect?

File: doc/sphinx/language/core/sections.rst [hunk 9]
```diff
@@ -131,6 +131,90 @@ usable outside the section as shown in this :ref:`example <section_local_declara
...
+    - in effect
+    - in effect
```
Comment (jfehrle) [inline doc/sphinx/language/core/sections.rst:181]: How is this different from being "available" or "supported"?

File: doc/sphinx/language/core/sections.rst [hunk 10]
```diff
@@ -131,6 +131,90 @@ usable outside the section as shown in this :ref:`example <section_local_declara
    Notice the difference between the value of :g:`x'` and :g:`x''` inside section
    :g:`s1` and outside.
 
+.. _visibility-attributes-sections:
+
+Summary of locality attributes in a section
+-------------------------------------------
+
+The following table sums up the locality of vernacular commands in sections, when
+outside the section where it was entered.
```
Comment (jfehrle) [inline doc/sphinx/language/core/sections.rst:140]:
```suggestion
Scope of locality attributes in a section
-----------------------------------------

This table summarizes the scope of the effects of vernacular commands outside
the section containing the command.
```

Better?  If you with this, you may want to tweak the wording used in table entries.  WDYT?

File: doc/sphinx/language/core/modules.rst [hunk 11]
```diff
@@ -511,6 +511,249 @@ transparent constraint
...
+    - same as :attr:`global`
+    - not available
+    - attribute not supported
+    - available when imported
+
+  * - :cmd:`Ltac`
+    - same as :attr:`global`
+    - not available
+    - attribute not supported
+    - available with short name if imported, with qualified name if not
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:737]: Let's hope this longish text gets line breaks in the rendered doc.  IIRC that may not be the case.  (Still waiting for the doc job to run.)

File: doc/sphinx/language/core/modules.rst [hunk 12]
```diff
@@ -511,6 +511,249 @@ transparent constraint
...
+
+.. _visibility-attributes-modules:
+
+Summary of locality attributes in a module
+------------------------------------------
+
+The following table sums up the locality of vernacular commands in modules, when
+outside the module where it was entered.
+A similar table for :cmd:`Section` can be found
+:ref:`here<visibility-attributes-sections>`.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:678]: Consider rewording similar to my comment on the other table.

File: doc/sphinx/language/core/sections.rst [hunk 13]
```diff
@@ -131,6 +131,90 @@ usable outside the section as shown in this :ref:`example <section_local_declara
...
+A similar table for :cmd:`Module` can be found
```
Comment (jfehrle) [inline doc/sphinx/language/core/sections.rst:141]: Perhaps you could include an asterisk on entries where the module and section behavior is different (in both tables)?  Won't users frequently want to compare the behavior of the two?

(I would have suggested making a single table that covers both sections and modules, but that likely won't fit the page width and may be hard to read.)

**Review by jfehrle — COMMENTED** [review]
wording/formatting suggestions

File: doc/sphinx/language/core/sections.rst (hunk 9 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/sections.rst:181]: Thank you for your comments. This is probably unclear. What I meant was that the setting itself is in effect outside the module or section.

File: doc/sphinx/language/core/sections.rst (hunk 10 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/sections.rst:140]: I'm not sure about this one. Scope would be the corresponding term in other programming languages, but in Coq it is really tight to notations and I don't want to create more confusion.

File: doc/sphinx/language/core/sections.rst (hunk 8 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/sections.rst:179]: Here, no effect means that the Setting has no effect outside of the current module or section.

File: doc/sphinx/language/core/sections.rst (hunk 7 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/sections.rst:156]: "not supported" means that there is an error message, yes. A ❌ could make it easier to read, you're right.

File: doc/sphinx/language/core/modules.rst (hunk 5 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst:514]: Alright, I can do that with a little bit of git struggle next weak.

File: doc/sphinx/language/core/sections.rst (hunk 10 above)
Comment (jfehrle) [reply doc/sphinx/language/core/sections.rst:140]: Many of the terms we use in Coq doc have multiple meanings depending on the context.  For example "scope" is already used in the programming language sense in the modules chapter for describing the locality attributes:

image

When necessary, we try to be more explicit in the doc, e.g. there are references to "notation scope", which is in the glossary BTW.

File: doc/sphinx/language/core/modules.rst (hunk 5 above)
Comment (jfehrle) [reply doc/sphinx/language/core/modules.rst:514]: Use `git rebase -i HEAD~999`, substituting something sensible for the `999` to move the text before your other changes.  Repeat moving the text and commit that before `git rebase continue`.  (Glossing over some details.  I'd be happy to be more explicit if you like.)

File: doc/sphinx/language/core/modules.rst (hunk 11 above)
Comment (jfehrle) [reply doc/sphinx/language/core/modules.rst:737]: Alas, the lines are not split.  The user would have to scroll the table right/left--much harder to read.  You might be able to fit it in the available width by much shorter mnemonics explained just above or below the main table.  They could be hyperlinked if you like.

image

File: doc/sphinx/language/core/sections.rst (hunk 13 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/sections.rst:146]: I think they are very different (and locality attributes are less available in sections), would you have suggestions?

**Comment by Villetaneuse** [comment]
@jfehrle Is this new version better?

File: doc/sphinx/language/core/sections.rst (hunk 9 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/sections.rst]: I have changed it to "available" for consistency with an explanation of what it means for this table.

File: doc/sphinx/language/core/sections.rst (hunk 10 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/sections.rst]: I have not changed the subtitle, but mentioned the scope in the explanation.

**Comment by jfehrle** [comment]
The formatting looks much better--looks like line wrapping works for cells in the table but not for the headers.  A few further suggestions shortly.

File: doc/sphinx/language/core/modules.rst (hunk 5 above)
Comment (jfehrle) [reply doc/sphinx/language/core/modules.rst:516]: A good try, but the "Move Typing Modules" commit also includes the added table.  If you want to give it another try or would like help, let me know.  It's not the end of the world if it remains as you have it now.

File: doc/sphinx/language/core/modules.rst [hunk 14]
```diff
@@ -511,6 +511,268 @@ transparent constraint
...
+------------------------------------------
+
+This table sums up the effect of locality attributes on the scope of vernacular
+commands in a module, when outside the module where they were entered. A cross
+(❌) marks an unsupported attribute, which will provoke a compilation error. In
+this table, "not available", means that the command has no effect outside the
+module it was entered and "short name when imported" means that the command
+always has effects outside the module but if the module (or the command, via
+selective importation, when available) is not imported, the corresponding
+identifier must be qualified in order to be used.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:684]:
```suggestion
this table, "not available" means that the command has no effect outside the
module containing it.  "Short name when imported" means that the command
always has effects outside the module; if the module is not imported, the
associated identifiers must be qualified.
```

"(or the command, via selective importation, when available)" - seemed less important than the other bits, maybe make it the final sentence?  OTOH, "when available" suggests that identifiers defined by some commands can be selectively imported and other cannot.  Would it be worthwhile to identify those differences?

File: doc/sphinx/language/core/modules.rst [hunk 15]
```diff
@@ -511,6 +511,268 @@ transparent constraint
...
+.. list-table::
+  :header-rows: 1
+
+  * - ``Command``
+    - no attribute
+    - :attr:`local`
+    - :attr:`export`
+    - :attr:`global`
+
+  * - :cmd:`Definition`, :cmd:`Lemma`,
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:698]: Is there a rationale for the order of the items?  For a short list, it doesn't have to be alphabetical.  It could be grouped by similar types of commands or perhaps by grouping commands with similar behavior.  Probably the two tables should have the identical order for ease of comparison.  Whatever you think is best.

File: doc/sphinx/language/core/sections.rst (hunk 13 above)
Comment (jfehrle) [reply doc/sphinx/language/core/sections.rst:146]: Now that I can see both tables, the section table data appears fairly simple/relatively consistent, so I wouldn't bother with asterisks.

**Review by jfehrle — COMMENTED** [review]
A few more suggestions.

Adding these tables is a helpful addition to the doc!

File: doc/sphinx/language/core/modules.rst [hunk 16]
```diff
@@ -511,6 +511,258 @@ transparent constraint
...
+  * - :cmd:`Set` or :cmd:`Unset` a flag
+    - :attr:`local`
+    - not available
+    - when imported
+    - always
+
+  * - :cmd:`Canonical Structure`
+    - :attr:`local`
+
+      or :attr:`global`
```
[the remark this answers is not in the document]
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst:737]: For this one, I just saw experimentally that default, local and global have the same meaning for the Canonical command. See also issue #19348. I did not manage to find the .ml file where this default locality attribute is fixed. @SkySkimmer would you know where to find this piece of information?

[the remark this answers is not in the document]
Comment (SkySkimmer) [reply doc/sphinx/language/core/modules.rst:737]: https://github.com/coq/coq/blob/e564bef2404475ec83061cf7bf4b0759d58aca85/vernac/canonical.ml#L30

`cache` is run when adding the object and at the end of sections if it's kept
`open` is run when imported, in this case it checks `Int.equal i 1` which means it does nothing for deep imports (eg if the object is from M.N when importing M)
`load` is run when requiring and at the end of modules, in this case it's not provided so it does nothing
`subst` is module substitution, doesn't matter for locality
`classify` tells if the object is kept at the end of modules, in this case it's Substitute so it's always kept. Dispose would mean it's dropped. There's also Keep which means it's kept but doesn't work with functors, and Anticipate which is only used for Require because it needs to coordinate with the kernel
`discharge` tells if the object is kept at the end of sections, in this case it depends on the locality boolean

So for the canonical structure object we see
- `local` is default false (unless the callers have some other default, in this case they don't)
- it is kept at the end of sections if not local and not a variable from the section, so eg `Section S. Variable x : ... Canonical Structure x. End S. (* not kept *)`
- it is always kept at the end of modules and takes effect on non-deep Import, not Require

See also https://github.com/coq/coq/blob/e564bef2404475ec83061cf7bf4b0759d58aca85/library/libobject.mli#L15-L65

**Comment by SkySkimmer** [comment]
I'm not sure about this approach with a centralized table. It feels like it's going to be more work to maintain than keeping the info in the doc for each command (eg https://coq.inria.fr/doc/master/refman/proofs/automatic-tactics/auto.html#creating-hints "Outside of sections, these commands support the local, export and global attributes. export is the default." for Hint commands)

**Comment by Villetaneuse** [comment]
> I'm not sure about this approach with a centralized table. It feels like it's going to be more work to maintain than keeping the info in the doc for each command (eg https://coq.inria.fr/doc/master/refman/proofs/automatic-tactics/auto.html#creating-hints "Outside of sections, these commands support the local, export and global attributes. export is the default." for Hint commands)

Don't you think the test files will make them easier to maintain?

**Comment by jfehrle** [comment]
> I'm not sure about this approach with a centralized table. It feels like it's going to be more work to maintain than keeping the info in the doc for each command

Yes, it would be best to include the info under each command description in addition to the table (perhaps hyperlinked).  Briefly repeating selected essential info is helpful to readers.  If @Villetaneuse is willing to check/update such descriptions, that would be great, but that shouldn't block this PR as the table is a useful improvement in itself.  If we're not expecting the info to change, I don't see a significant maintenance burden.  The maintenance would not be difficult, either.

**Comment by thomas-lamiaux** [comment]
Should include be discussed there too ? 
https://stackoverflow.com/questions/78803435/coqs-search-command-and-modules

**Comment by SkySkimmer** [comment]
That depends what you call copy paste, eg Include won't include Local effects.

**Comment by Villetaneuse** [comment]
> That depends what you call copy paste, eg Include won't include Local effects.

Thank you very much for this piece of knowledge. But... this means I really have to also check what happens with `Include` too?

**Comment by SkySkimmer** [comment]
Include is like Require + Export

**Comment by SkySkimmer** [comment]
To be more precise it's "rename objects to be from the current module" then Require + Export the renamed objects

File: doc/sphinx/language/core/modules.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst:737]: Thank you very much for this very informative answer!

I'll settle for "local", then, it's consistent with what is in sections and closer to what happens (the Canonical is available outside the module by default).

File: doc/sphinx/language/core/modules.rst (hunk 14 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst:684]: You're right, I'm dropping "when available" since it adds some noise in an already complicated state.
I think the refman is already clear about selective importation, but I will see if it can be improved later.

File: doc/sphinx/language/core/modules.rst (hunk 15 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst:698]: I have changed the order to group commands by identical behaviour.

**Comment by Villetaneuse** [comment]
> To be more precise it's "rename objects to be from the current module" then Require + Export the renamed objects

Then, I don't think it should be mentioned with these tables, but rather in the part of the refman about `Include`.

**Comment by Villetaneuse** [comment]
I have taken most comments into account. I think this leads to clearer documentation.
Some issues have been raised during review:
[ ] what about `Include`
  I suggest a separate PR to clarify the behaviour of `Include` in the refman
[ ] ease of maintenance
  I suggest taking a look at the documentation of each command and provide links to each documentation in the associated test files
[x] surprising entries: the only one which comes to mind is `Canonical` in modules, there is an associated issue: #19348

File: doc/sphinx/language/core/modules.rst (hunk 5 above)
Comment (jfehrle) [reply doc/sphinx/language/core/modules.rst:520]: Don't forget about this one (make "[Move "Typing modules" to the end](https://github.com/coq/coq/pull/19323/commits/9710d79eefed34a9c0465c1fc77fe1c7b949b409)" only moving text, not adding the table).

File: doc/sphinx/language/core/modules.rst [hunk 17]
```diff
@@ -515,6 +515,274 @@ transparent constraint
...
+This table sums up the effect of locality attributes on the scope of vernacular
+commands in a module, when outside the module where they were entered. In the
+following table:
+
+* a cross (❌) marks an unsupported attribute (compilation error);
+* "not available" means that the command has no effect outside the module it was entered;
+* "when imported" means that the command has effect outside the module if, and
+  only if, the module (or the command, via selective importation) is imported;
+* "short name when imported" means that the command has effects outside the module;
+  if the module (or command, via selective importation) is not imported, the
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:690]:
```suggestion
* "short name when imported" means that the command has effects outside the module;
  if the module (or command, via :n:`@filtered_import`) is not imported, the
```

"selective import" appears only one place in the existing doc and is inconsistent with the nonterminal "filtered_import" used in Import and Require commands

File: doc/sphinx/language/core/modules.rst [hunk 18]
```diff
@@ -515,6 +515,274 @@ transparent constraint
...
+------------------------------------------
+
+This table sums up the effect of locality attributes on the scope of vernacular
+commands in a module, when outside the module where they were entered. In the
+following table:
+
+* a cross (❌) marks an unsupported attribute (compilation error);
+* "not available" means that the command has no effect outside the module it was entered;
+* "when imported" means that the command has effect outside the module if, and
+  only if, the module (or the command, via selective importation) is imported;
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:688]:
```suggestion
* "when imported" means that the command has effect outside the module if, and
  only if, the module (or the command, via selective importation) is imported (with 
  :cmd:`Import`, :cmd:`Export` or :cmd:`Require`);
```

If you're willing, maybe clarify the difference between Import and Export under those commands.  The doc for Export just says "Similar to Import, except that when the module containing this command is imported, the qualid+ are imported as well.", which only tells me "something is different between Import and Export--guess what it is?"

**Review by jfehrle — COMMENTED** [review]
A few further comments.  This is looking pretty good.  It doesn't have to be absolutely perfect as long as it's a genuine improvement (helpful to users).

File: test-suite/success/locality_attributes_modules.v [hunk 19]
```diff
@@ -0,0 +1,402 @@
+(** This file tests how locality attributes affect usual vernacular commands.
+    PLEASE, when this file fails to compute following a voluntary change in
+    Coq's behaviour, modify accordingly the tables in [sections.rst] and
+    [modules.rst] in [doc/sphinx/language/core].
+
+    Also look at the corresponding discussions about locality attributes in the
+    refman (directory doc/sphinx)
+    - For Definition, Lemma, ..., look at language/core/definitions.rst
+    - For Axiom, Conjecture, ..., look at language/core/assumptions.rst
+    - For abbreviations, look at user-extensions/syntax-extensions.rst
```
Comment (jfehrle) [inline test-suite/success/locality_attributes_modules.v:10]: This may be more detail than necessary.  It's easy to find these in the Commands index and should be obvious to any Coq developer.  (To find "abbreviation", do a text search for "abbreviation" within that index).  Your choice.

Comment (Villetaneuse) [reply test-suite/success/locality_attributes_modules.v:10]: I don't think it really hurts, unless the refman's files are reorganized, which is possible but (I guess) less likely than a change in the locality attributes behaviour. It's a strong incentive for correcting the refman if such a change happens.

**Comment by Villetaneuse** [comment]
I think I'm done with this PR. (I have noted as TODO to work on `Export` and `Include` documention).

File: doc/sphinx/language/core/modules.rst (hunk 18 above)
Comment (Villetaneuse) [reply doc/sphinx/language/core/modules.rst]: done (without `Require` which leaves in the file world).

**Review by jfehrle — APPROVED** [review]
Looks good, thanks for the PR!

@coqbot: merge now

**Comment by jfehrle** [comment]
@coqbot: merge now
